Add new test_plugin for openstack preflight

hemna opened the pull request:
This patch adds a new test_plugins directory
    and a test plugin called openstack-preflight.py
    that adds some test for making sure images, networks
    security groups are available against the configured
    openstack deployment.

    the test plugin adds the following tests
    os_check_image - check to see if an image is available on OpenStack
    os_check_network - check to see if a network is defined on OpenStack
    os_check_subnet - check to see if an subnet is defined on OpenStack
    os_check_security_group - check to see if a security group is defined on OpenStack

    Example usage:

    - name: test to make sure my image is available
      fail:
       msg: "Failed to find {{my_image}} on OpenStack
      when: not my_image is os_check_image

    This also adds ansible native tests for caasp, ses, osh
    deployments to ensure their required images and security groups
    are in place.  The failure messages provide detail about what's
    missing.

jesusaurus commented on script_library/openstack-preflight.py:
```diff
@@ -0,0 +1,158 @@
+                    help="Run all the checks?",
+                    default=False, action="store_true")
+
+CLOUD = os.getenv('OS_TEST_CLOUD', 'devstack')
```
Could we use OS_CLOUD here instead of creating a new variable?

hemna replied:
done

hemna commented:
Hi JP,
  I added this test because I was struggling to find out the required images, networking options, etc that are placed in the vars/deploy-on-openstack.yml file.  If you try and run any of the deploy_* commands with run.sh you will run into a ton of missing items in my own openstack deployment, that I then have to track down and spend a ton of time doing.   Ansible simply spits out errors with large stack traces that are impossible to read to find the real error, a missing image named X.  This wastes a ton of time, and this simple python script tests and spits out errors that are human readable and easy to rectify in a short period of time.   We desperately need some checks in place like this to make this usable to a wider audience.

evrardjp commented:
> Hi JP,
> I added this test because I was struggling to find out the required images, networking options, etc that are placed in the vars/deploy-on-openstack.yml file. If you try and run any of the deploy_* commands with run.sh you will run into a ton of missing items in my own openstack deployment, that I then have to track down and spend a ton of time doing. Ansible simply spits out errors with large stack traces that are impossible to read to find the real error, a missing image named X. This wastes a ton of time, and this simple python script tests and spits out errors that are human readable and easy to rectify in a short period of time. We desperately need some checks in place like this to make this usable to a wider audience.

This should be all documented already in the deploy guide. It means we have a gap.
Are you trying to run on your own openstack cloud?

hemna commented:
> > Hi JP,
> > I added this test because I was struggling to find out the required images, networking options, etc that are placed in the vars/deploy-on-openstack.yml file. If you try and run any of the deploy_* commands with run.sh you will run into a ton of missing items in my own openstack deployment, that I then have to track down and spend a ton of time doing. Ansible simply spits out errors with large stack traces that are impossible to read to find the real error, a missing image named X. This wastes a ton of time, and this simple python script tests and spits out errors that are human readable and easy to rectify in a short period of time. We desperately need some checks in place like this to make this usable to a wider audience.
> 
> This should be all documented already in the deploy guide. It means we have a gap.
> Are you trying to run on your own openstack cloud?
> 
> Next to that, ARA would help on the debugging if you don't want to browse your scrollback. It's also documented in the deploy guide.

It's not documented entirely.   Yes I'm trying to deploy against my own openstack deployment.  I think we should have the checks in place to help out.   Would you prefer I make an ansible plugin to allow us to check these entries on the openstack deployment?   The current failures w/o this script in place is very difficult to find the actual missing requirements.   Hence the need for all the checks I added in the script.

toabctl commented on playbooks/openstack-caasp_instance.yml:
```diff
@@ -10,6 +10,9 @@
       include_vars: "{{ playbook_dir }}/../vars/deploy-on-openstack.yml"
 
+    - name: Preflight OpenStack Caasp resources
+      include: _openstack_preflight_caasp.yml
```
This gets now called whenever we call the deploy caasp step. Even if the stack got already successfully deployed. I don't think that is needed and wasted time

hemna replied:
refactored

toabctl commented on playbooks/openstack-osh_instance.yml:
```diff
@@ -8,6 +8,10 @@
         - "{{ playbook_dir }}/../vars/deploy-on-openstack.yml"
+
+    - name: Preflight OpenStack oshnode resources
+      include: _openstack_preflight_osh.yml
```
same here (gets always called)

hemna replied:
refactored

toabctl commented on playbooks/openstack-ses_aio_instance.yml:
```diff
@@ -8,6 +8,10 @@
         - "{{ playbook_dir }}/../vars/deploy-on-openstack.yml"
+
+    - name: Preflight OpenStack SES resources
+      include: _openstack_preflight_ses.yml
```
same here

hemna replied:
refactored

toabctl commented on playbooks/test_plugins/openstack-preflight.py:
```diff
@@ -0,0 +1,113 @@
+    return found
+
+
+#def ses_checks(cloud, os_vars):
```
Can you remove the commented code?

hemna replied:
done

hemna commented:
> In general I would prefer to have these kind of checks in an extra step (so it is not called every time). Then we can document that a user can run the check when needed.

Currently everytime you execute run.sh deploy_caasp all of the preflight checks are run as well.  Shall we remove those as well?

rhafer commented on playbooks/openstack-preflight_osh.yml:
```diff
@@ -0,0 +1,28 @@
+# Playbook to verify that required resources are in the OpenStack
+# Cloud
+
+---
```
This doesn't seem to be a valid ansible play.  It's missing the toplevel attributes. (See also the test failure https://ci.suse.de/blue/organizations/jenkins/cloud-socok8s%2Fsocok8s/detail/PR-270/5/pipeline/34/#step-43-log-132)

Same is true for `openstack-preflight_ses.yml`

hemna replied:
fixed both of those.

dirkmueller commented:
@hemna rebase?
